File: java/examples/src/main/java/com/optikpi/examples/TestCustomerEndpoint.java
package com.optikpi.examples;

import java.time.Instant;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

import com.optikpi.datapipeline.ClientConfig;
import com.optikpi.datapipeline.OptikpiDataPipelineSDK;
import com.optikpi.datapipeline.model.CustomerProfile;
import com.optikpi.datapipeline.model.ValidationResult;

import io.github.cdimascio.dotenv.Dotenv;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;

/**
 * Example: Test Customer Profile Endpoint Demonstrates how to send customer
 * profile data to the Optikpi Data Pipeline API
 */
public class TestCustomerEndpoint {

    private static final ObjectMapper mapper = new ObjectMapper()
            .enable(SerializationFeature.INDENT_OUTPUT);

    public static void main(String[] args) {
        // Load environment variables
         Dotenv dotenv = Dotenv.configure()
                .directory(".")
                .filename(".env")
                .load();

        // Get configuration from environment variables
        String authToken = dotenv.get("AUTH_TOKEN");
        String accountId = dotenv.get("ACCOUNT_ID");
        String workspaceId = dotenv.get("WORKSPACE_ID");
        String baseUrl = dotenv.get("API_BASE_URL");

        if (authToken == null || accountId == null || workspaceId == null) {
            System.err.println("Error: Missing required environment variables:");
            System.err.println("Please set OPTIKPI_AUTH_TOKEN, OPTIKPI_ACCOUNT_ID, and OPTIKPI_WORKSPACE_ID");
            System.exit(1);
        }

        // Create client configuration
        ClientConfig config = new ClientConfig(authToken, accountId, workspaceId);
        config.setBaseUrl(baseUrl);

        // Create SDK instance
        OptikpiDataPipelineSDK sdk = new OptikpiDataPipelineSDK(config);

        System.out.println("🚀 Testing Customer Profile Endpoints");
        System.out.println("======================================");
        System.out.println("Configuration:");
        System.out.println("📌 API Base URL: " +  config.getBaseUrl());
        System.out.println("👤 Account ID: " + config.getAccountId());
        System.out.println("🏢 Workspace ID: " + config.getWorkspaceId());
        System.out.println();

        testCustomerEndpoint(sdk, accountId, workspaceId);
    }

    private static void testCustomerEndpoint(OptikpiDataPipelineSDK sdk, String accountId, String workspaceId) {
        try {
            // Create customer data matching JavaScript example
            CustomerProfile customer = new CustomerProfile();
            customer.setAccountId(accountId);
            customer.setWorkspaceId(workspaceId);
            customer.setUserId("java_t_03");
            customer.setUsername("john_doe");
            customer.setFullName("John Doe");
            customer.setFirstName("John");
            customer.setLastName("Doe");
            customer.setDateOfBirth("1990-01-15");
            customer.setEmail("john.doe@example.com");
            customer.setPhoneNumber("+1234567890");
            customer.setGender("Male");
            customer.setCountry("United States");
            customer.setCity("New York");
            customer.setLanguage("en");
            customer.setCurrency("USD");

            customer.setPhoneVerification("Verified");
            customer.setEmailVerification("Verified");
            customer.setBankVerification("NotVerified");
            customer.setIddocVerification("Verified");
        
            customer.setMarketingEmailPreference("Opt-in");
            customer.setNotificationsPreference("Opt-in");
            customer.setSubscription("Subscribed");
    
            customer.setPrivacySettings("private");
            
            customer.setDepositLimits(1000.00);
            customer.setLossLimits(500.00);
            customer.setWageringLimits(2000.00);
            customer.setSessionTimeLimits(120);

            customer.setCoolingOffExpiryDate("2024-12-31T23:59:59Z");
            customer.setSelfExclusionExpiryDate("2025-01-31T23:59:59Z");
            customer.setAffliateId("affiliate456");
            
            customer.setCoolingOffPeriod(7);
            customer.setSelfExclusionPeriod(30);

            customer.setRealityChecksNotification("daily");
            customer.setVipStatus("Regular");
            customer.setLoyaltyProgramTiers("Bronze");
            customer.setAccountStatus("Active");
            customer.setBonusAbuser("Not flagged");
            customer.setFinancialRiskLevel(0.3);
            
            customer.setAcquisitionSource("Google Ads");
            customer.setPartnerId("partner123");
            customer.setReferralLinkCode("REF789");
            customer.setReferralLimitReached("Not Reached");
            customer.setCreationTimestamp("2024-01-15T10:30:00Z");
        
            customer.setRiskScoreLevel("low");
            customer.setMarketingSmsPreference("Opt-in");
            
            // Custom data as Map (same in both)
            Map<String, Object> customData = new HashMap<>();
            customData.put("favorite_game", "slots");
            customData.put("newsletter_signup", true);
            customer.setCustomData(customData);
            
            customer.setSelfExclusionBy("player");
            customer.setSelfExclusionByType("voluntary");
            customer.setSelfExclusionCheckTime("2024-01-15T10:30:00Z");
            customer.setSelfExclusionCreatedTime("2024-01-01T00:00:00Z");
            customer.setClosedTime(null);

            customer.setRealMoneyEnabled("true");
            customer.setPushToken("push_token_abc123");
            customer.setAndroidPushToken("bRdkk1jSTdKFxOkO3c26rt:APA91bEoaBQn0TnxH7gqJJtyzD9cwUQBGJOtVhQ8kJv6v52Opqb2CTUlfFnPxSPM1PSSG9Ctlb1ASuVfPvKCTchaz_LzZH03TuigP4TsYKYZ76FEKh_GZoA");
            customer.setIosPushToken("bRdkk1jSTdKFxOkO3c26rt:APA91bEoaBQn0TnxH7gqJJtyzD9cwUQBGJOtVhQ8kJv6v52Opqb2CTUlfFnPxSPM1PSSG9Ctlb1ASuVfPvKCTchaz_LzZH03TuigP4TsYKYZ76FEKh_GZoA");
            customer.setWindowsPushToken("bRdkk1jSTdKFxOkO3c26rt:APA91bEoaBQn0TnxH7gqJJtyzD9cwUQBGJOtVhQ8kJv6v52Opqb2CTUlfFnPxSPM1PSSG9Ctlb1ASuVfPvKCTchaz_LzZH03TuigP4TsYKYZ76FEKh_GZoA");
            customer.setMacdmgPushToken("bRdkk1jSTdKFxOkO3c26rt:APA91bEoaBQn0TnxH7gqJJtyzD9cwUQBGJOtVhQ8kJv6v52Opqb2CTUlfFnPxSPM1PSSG9Ctlb1ASuVfPvKCTchaz_LzZH03TuigP4TsYKYZ76FEKh_GZoA");

            // Validate customer data
            ValidationResult valid = customer.validate();
            if (!valid.isValid()) {
                System.out.println("\n❌ Validation Failed!");
                System.out.println("Errors: " + valid.getErrors());
                return;
            }
            
            System.out.println("✅ Customer event validated successfully!");
            System.out.println("Customer Data:");
            System.out.println(mapper.writeValueAsString(customer));

            System.out.println("\n🕒 Making API request using SDK...");
            long start = System.currentTimeMillis();
            var response = sdk.sendCustomerProfile(customer);
            long end = System.currentTimeMillis();

            System.out.println("\n📡 API Response:");
            System.out.println("⏱ Response Time: " + (end - start) + "ms");
            System.out.println("HTTP Status: " + response.getStatus());
            System.out.println("Sending Customer Profile event...");

            if (response.isSuccess()) {
                System.out.println("✅ SUCCESS!");
                System.out.println("Response: " +
                        mapper.writeValueAsString(response.getData()));
            } else {
                System.out.println("❌ FAILED!");
                System.out.println("Error: " +
                        mapper.writeValueAsString(response.getError()));
            }

        } catch (Exception e) {
            System.err.println("\n💥 Exception occurred:");
            e.printStackTrace();
        }
    }
}

File: php/examples/test-customer-endpoint.php
<?php

require_once __DIR__ . '/../vendor/autoload.php';

use Optikpi\DataPipeline\OptikpiDataPipelineSDK;
use Optikpi\DataPipeline\Models\CustomerProfile;

// Load environment variables
$dotenv = parse_ini_file(__DIR__ . '/../.env');

$API_BASE_URL = $dotenv['API_BASE_URL'] ?? getenv('API_BASE_URL');
$AUTH_TOKEN = $dotenv['AUTH_TOKEN'] ?? getenv('AUTH_TOKEN');
$ACCOUNT_ID = $dotenv['ACCOUNT_ID'] ?? getenv('ACCOUNT_ID');
$WORKSPACE_ID = $dotenv['WORKSPACE_ID'] ?? getenv('WORKSPACE_ID');

// Validate required environment variables
if (empty($AUTH_TOKEN) || empty($ACCOUNT_ID) || empty($WORKSPACE_ID)) {
    echo "❌ Error: Missing required environment variables!\n";
    echo "   Please set: AUTH_TOKEN, ACCOUNT_ID, WORKSPACE_ID\n";
    echo "   Copy env.example to .env and fill in your values\n";
    exit(1);
}

// Initialize SDK
$sdk = new OptikpiDataPipelineSDK([
    'authToken' => $AUTH_TOKEN,
    'accountId' => $ACCOUNT_ID,
    'workspaceId' => $WORKSPACE_ID,
    'baseURL' => $API_BASE_URL
]);

// Customer data
$customer = new CustomerProfile([
    'account_id' => $ACCOUNT_ID,
    'workspace_id' => $WORKSPACE_ID,
    'user_id' => 'js_field04prod',
    'username' => 'john_doe',
    'full_name' => 'John Doe',
    'first_name' => 'John',
    'last_name' => 'Doe',
    'date_of_birth' => '1990-01-15',
    'email' => 'john.doe@example.com',
    'phone_number' => '+1234567890',
    'gender' => 'Male',
    'country' => 'United States',
    'city' => 'New York',
    'language' => 'en',
    'currency' => 'USD',
    'marketing_email_preference' => 'Opt-in',
    'notifications_preference' => 'Opt-in',
    'subscription' => 'Subscribed',
    'privacy_settings' => 'private',
    'deposit_limits' => 1000.00,
    'loss_limits' => 500.00,
    'wagering_limits' => 2000.00,
    'session_time_limits' => 120,
    'reality_checks_notification' => 'daily',
    'account_status' => 'Active',
    'vip_status' => 'Regular',
    'loyalty_program_tiers' => 'Bronze',
    'bonus_abuser' => 'Not flagged',
    'financial_risk_level' => 0.3,
    'acquisition_source' => 'Google Ads',
    'partner_id' => 'partner123',
    'referral_link_code' => 'REF789',
    'referral_limit_reached' => 'Not Reached',
    'creation_timestamp' => '2024-01-15T10:30:00Z',
    'phone_verification' => 'Verified',
    'email_verification' => 'Verified',
    'bank_verification' => 'NotVerified',
    'iddoc_verification' => 'Verified',
    'cooling_off_expiry_date' => '2024-12-31T23:59:59Z',
    'self_exclusion_expiry_date' => '2025-01-31T23:59:59Z',
    'risk_score_level' => 'low',
    'marketing_sms_preference' => 'Opt-in',
    'custom_data' => [
        'favorite_game' => 'slots',
        'newsletter_signup' => true
    ],
    'self_exclusion_by' => 'player',
    'self_exclusion_by_type' => 'voluntary',
    'self_exclusion_check_time' => '2024-01-15T10:30:00Z',
    'self_exclusion_created_time' => '2024-01-01T00:00:00Z',
    'closed_time' => null,
    'real_money_enabled' => 'true',
    'push_token' => 'push_token_abc123',
    'android_push_token' => 'android_push_token_xyz456',
    'ios_push_token' => 'ios_push_token_def789',
    'windows_push_token' => 'windows_push_token_ghi012',
    'mac_dmg_push_token' => 'mac_push_token_jkl345'
]);

$validation = $customer->validate();
if (!$validation['isValid']) {
    echo "❌ Validation errors:\n";
    foreach ($validation['errors'] as $error) {
        echo "  - $error\n";
    }
    exit(1);
}
echo "✅ Customer event validated successfully!\n";

// SDK handles HMAC authentication automatically

// Test customer endpoint
function testCustomerEndpoint($sdk, $customer, $API_BASE_URL, $ACCOUNT_ID, $WORKSPACE_ID, $AUTH_TOKEN) {
    try {
        echo "🚀 Testing Customer Endpoint\n";
        echo "============================\n";

        echo "Configuration:\n";
        echo "API Base URL: $API_BASE_URL\n";
        echo "Account ID: $ACCOUNT_ID\n";
        echo "Workspace ID: $WORKSPACE_ID\n";
        echo "Auth Token: " . substr($AUTH_TOKEN, 0, 8) . "...\n";

        echo "\nMaking API request using SDK...\n";
        echo "Customer Data: " . json_encode($customer->toArray(), JSON_PRETTY_PRINT) . "\n";

        // Make the API call using SDK
        $startTime = microtime(true) * 1000;
        $result = $sdk->sendCustomerProfile($customer);
        $endTime = microtime(true) * 1000;

        if ($result['success']) {
            echo "\n✅ Success!\n";
            echo "============================\n";
            echo "HTTP Status: " . $result['status'] . "\n";
            echo "Response Time: " . round($endTime - $startTime) . "ms\n";
            echo "SDK Success: " . ($result['success'] ? 'true' : 'false') . "\n";
            echo "Response Data: " . json_encode($result['data'], JSON_PRETTY_PRINT) . "\n";
        } else {
            echo "\n❌ API Error!\n";
            echo "============================\n";
            echo "HTTP Status: " . ($result['status'] ?? 'N/A') . "\n";
            echo "Response Time: " . round($endTime - $startTime) . "ms\n";
            echo "SDK Success: " . ($result['success'] ? 'true' : 'false') . "\n";
            echo "Error Data: " . json_encode($result['data'], JSON_PRETTY_PRINT) . "\n";
        }

    } catch (Exception $error) {
        echo "\n❌ SDK Error occurred!\n";
        echo "============================\n";
        echo "Error: " . $error->getMessage() . "\n";
        echo "Stack: " . $error->getTraceAsString() . "\n";
    }
}

// Run the test
if (php_sapi_name() === 'cli') {
    testCustomerEndpoint($sdk, $customer, $API_BASE_URL, $ACCOUNT_ID, $WORKSPACE_ID, $AUTH_TOKEN);
}

